### PR TITLE
Upgrade blockstore to use Python 3.8 and Django 2.2+.

### PR DESCRIPTION
**Pull request by Kelketek**
## Description

This pull request upgrades Blockstore to use Python 3.8. It also upgrades its Django version to match that of the LMS, since the Django version we had does not officially support 3.8 and I found this to have a handful of strange side effects.

## Author Comments, Concerns, and Open Questions

I originally changed the docker-compose files to no longer use a cached volume for the venv, but after finding that it did, in fact, have pretty serious consequences on startup that it didn't seem to have when I checked initially, I decided to just change the volume name so that a new volume will be created with a new copy of Python.

I'd considered updating the docker file to just make installing those packages as part of the build, but with as many other things as this PR was touching, I felt a lighter change was a better idea.

This PR replaces https://github.com/edx/blockstore/pull/77 and https://github.com/edx/blockstore/pull/79, both of which have partial solutions. I've also added in some text to the readme to make it easier for newly onboarded devs to figure out how to run tests.

## Test Instructions

These are now in the Readme.

## Reviews
- [ ] @bradenmacdonald

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @Kelketek! I've created [OSPR-4961](https://openedx.atlassian.net/browse/OSPR-4961) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/tests/test_contract.py:321]: Stray `print` ?

File: blockstore/apps/bundles/tests/storage_utils.py
Comment (bradenmacdonald) [inline blockstore/apps/bundles/tests/storage_utils.py:115]: More `print`

**Comment by bradenmacdonald** [comment]
Triggered a CI build; apparently the commit has to be pushed to the `edx` repo.

**Comment by bradenmacdonald** [comment]
@Kelketek We can probably remove the python 3.5 tests as part of this PR.

@mraarif Do we need to merge https://github.com/edx/configuration/pull/5913 at the same time as this?

**Comment by mraarif** [comment]
you guys can plan when to merge the [configuration changes](https://github.com/edx/configuration/pull/5913), this PR can be merged independently.

**Comment by natabene** [comment]
@Kelketek Thank you for your contribution. Please let me know once it is ready for our review.

**Comment by Kelketek** [comment]
If @bradenmacdonald gives the go-ahead, I think we're ready. Braden, I've just resolved the quality issues, and as you mentioned, we can cut the 3.5 tests after this merge. Since @mraarif has indicated we can merge this independently, are we good to go from an internal review perspective?

**Comment by bradenmacdonald** [comment]
@Kelketek We should remove the 3.5 tests as part of this PR, so it shows a green build before we merge. Please include a commit to change `.circleci/config.yml`, removing the first occurrence of `3.5` and changing the rest to `3.8`

**Comment by Kelketek** [comment]
@bradenmacdonald Ah, right-- I blanked on being able to change that from the repository since I expected to see `circleci.yml` or similar in the root and it was in a directory at the top of my listing instead. I've updated the file but can't seem to trigger the tests on my own.

I've also gone ahead and squashed since there were few comments so far and they've already been resolved.

File: blockstore/apps/bundles/tests/factories.py
Comment (morenol) [inline blockstore/apps/bundles/tests/factories.py:3]: I think that the error here is due to:

 https://github.com/FactoryBoy/factory_boy/issues/780

**Comment by eLRuLL** [comment]
@Kelketek upgrading django comes with a migration to the User model (because of using the generic AbstractUser) changing the `last_name`'`max_length`, please run `makemigrations` to add it.

**Comment by Kelketek** [comment]
@eLRuLL I've added the migration. I've also fixed the merge conflict with the README and rewrote the part I added in reStructuredText instead of Markdown, since I discovered the README was .rst. Are we good to go here? As a reminder I'll need you or @bradenmacdonald to merge if we're clear to do so, since I don't have perms to push to this repository.

**Comment by eLRuLL** [comment]
LGTM 👍

**Comment by Kelketek** [comment]
@morenol I have updated the config to hopefully do so, but I can't trigger the tests to verify. I also discovered that there was a new conflict in a requirements file, which has been resolved.

**Comment by morenol** [comment]
@Kelketek we recently merged a PR that adds [OEP-18 compliance](https://open-edx-proposals.readthedocs.io/en/latest/oep-0018-bp-python-dependencies.html) to this repo. I think that you should remove the django2.2 and python3.5 related constraints from requirements/constraints.txt and then commit the changes produced by `make upgrade`

**Comment by Kelketek** [comment]
@morenol Done. :) There was not a 'python3.5' constraint-- that seems to be bound by the executable doing the upgrade, but there was a Django one and I've updated that.

File: .circleci/config.yml
Comment (bradenmacdonald) [reply .circleci/config.yml:17]: @Kelketek This is done right?

Comment (Kelketek) [reply .circleci/config.yml:17]: @bradenmacdonald Yep.

**Review by bradenmacdonald — APPROVED** [review]
:+1:

- [x] I tested this: ran the blockstore integration tests on my Open edX devstack
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

Note:

In testing this, I was getting a strange error:

```
Collecting inflect==4.1.0 (from -r requirements/local.txt (line 43))
  Could not find a version that satisfies the requirement inflect==4.1.0 (from -r requirements/local.txt (line 43)) (from versions: 0.1, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5, 0.1.6, 0.1.7, 0.1.8, 0.2.0, 0.2.1, 0.2.3, 0.2.4, 0.2.5, 0.3.0, 0.3.1, 1.0.0, 1.0.1, 1.0.2, 2.0.0, 2.0.1, 2.1.0, 3.0.1, 3.0.2)
No matching distribution found for inflect==4.1.0 (from -r requirements/local.txt (line 43))
```

(because docker-compose was caching my older image with python 3.5 and `inflect >4` is restricted to python 3.6+) The solution was to run `docker-compose -f docker-compose-testserver.yml build` and `docker-compose build`. Though I also had to manually delete `blockstore-testserver_blockstore_venv_3_8_5` because it had been provisioned using python 3.5 :/

**Comment by bradenmacdonald** [comment]
@kdmccormick Could you please hit merge on this as well? I've reviewed it as a core committer.

**Comment by kdmccormick** [comment]
@bradenmacdonald Can do. Are you alright with this being deployed tomorrow morning?

**Comment by bradenmacdonald** [comment]
@kdmccormick Yes, that's better for me too anyways :)

**Comment by edx-webhook** [comment]
@Kelketek 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/blockstore&entry.867055334=https://github.com/edx/blockstore/pull/82&entry.1484655318=https://github.com/Kelketek&entry.752974735=2020-09-10+21:10&entry.1917517419=2020-09-15+17:15&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by kdmccormick** [comment]
@bradenmacdonald @Kelketek Going to production with this now. I'll monitor on my side. Ping me on Slack if you see any issues on your side.

**Comment by kdmccormick** [comment]
@bradenmacdonald @Kelketek False alarm. We have some Blockstore deployment pipeline issues that we need to resolve before pushing this to prod. I'll update here once they are resolved and we can try again. Sorry all!

**Comment by kdmccormick** [comment]
Hey @Kelketek , in addition to the deployment pipeline weirdness, I decided to revert this because it does not maintain Python 3.5 compatibility. As long as this service's Python version is controlled independently via `edx/configuration`, all Python upgrades are best done with a period of overlap. This gives us more flexibility when rolling out, and, if necessary, rolling back. So the desired process here is:
1. Merge a Blockstore PR where tests pass in both 3.5 and 3.8. Code goes to Stage.
2. Smoke test Stage, which is running 3.5
3. Deploy to & smoke test Prod, which is running 3.5.
4. Merge an edx/configuration PR to switch over to 3.8.
5. Re-deploy Stage (with 3.8 this time) and smoke test.
7. Deploy to Prod (with 3.8 this time) and smoke test.
7a. If bugs arise, then revert to 3.5 via an edx/configuration PR, and return to step 1.
8. Merge Blockstore PR to remove support & tests for 3.5.


I should have realized this before merging the first time. My bad!

**Pull request merged**
